processor/src/fast/mod.rs

```diff
@@ -320,17 +323,86 @@ impl FastProcessor {
+                        MastNode::Block(basic_block_node) => {
+                            self.execute_basic_block_node(basic_block_node, &current_forest, host)
+                                .await?
+                        },
+                        MastNode::Join(join_node) => self.execute_join_node(
+                            join_node,
+                            current_node_id,
+                            phase,
+                            &mut continuation_stack,
+                        )?,
+                        MastNode::Split(split_node) => self.execute_split_node(
+                            split_node,
+                            current_node_id,
+                            phase,
+                            &current_forest,
+                            &mut continuation_stack,
+                        )?,
+                        MastNode::Loop(loop_node) => self.execute_loop_node(
+                            loop_node,
+                            current_node_id,
+                            phase,
+                            &current_forest,
+                            &mut continuation_stack,
+                        )?,
+                        MastNode::Call(call_node) => self.execute_call_node(
+                            call_node,
+                            current_node_id,
+                            phase,
+                            program,
+                            &current_forest,
+                            &mut continuation_stack,
+                        )?,
+                        MastNode::Dyn(dyn_node) => {
+                            self.execute_dyn_node(
+                                dyn_node,
+                                current_node_id,
+                                phase,
+                                &current_forest,
+                                &mut continuation_stack,
+                                host,
+                            )
+                            .await?
+                        },
+                        MastNode::External(external_node) => {
+                            self.execute_external_node(external_node, &mut continuation_stack, host)
+                                .await?
+                        },
+                    }
+                },
+                ProcessingStep::PreDecorators(node_id) => {
+                    let node = current_forest.get_node_by_id(node_id).unwrap();
+                    for &decorator_id in node.before_enter() {
+                        self.execute_decorator(&current_forest[decorator_id], 0, host)?;
+                    }
+                },
+                ProcessingStep::PostDecorators(node_id) => {
+                    let node = current_forest.get_node_by_id(node_id).unwrap();
+                    for &decorator_id in node.after_exit() {
+                        self.execute_decorator(&current_forest[decorator_id], 0, host)?;
+                    }
+                },
```

Comment:
Related to the previous comment, I would probably handle the decorators in node processing functions. It will add a bit more boilerplate code, but should also simplify continuation stack processing - i.e., the `ProcessingStep` could look something like this:
```rust
pub enum ProcessingStep{
    Start(MastNodeId),
    Finish(MastNodeId),
}
```
We could also consider making the `Finish` variant more explicit (because we always know what type of a node we are finishing) - e.g.:
```rust
pub enum ProcessingStep{
    Start(MastNodeId),
    FinishJoin, // don't care about node ID
    FinishSplit, // don't care about node ID
    FinishLoop(MastNodeId),
    ...
}
```

Reply:
Done